File: modules/ar_instruction_evaluator.c
/**
 * @file ar_instruction_evaluator.c
 * @brief Implementation of the instruction evaluator module
 */

#include "ar_instruction_evaluator.h"
#include "ar_heap.h"
#include "ar_assignment_instruction_evaluator.h"
#include "ar_send_instruction_evaluator.h"
#include "ar_condition_instruction_evaluator.h"
#include "ar_parse_instruction_evaluator.h"
#include "ar_build_instruction_evaluator.h"
#include "ar_method_instruction_evaluator.h"
#include "ar_agent_instruction_evaluator.h"
#include "ar_destroy_agent_instruction_evaluator.h"
#include "ar_destroy_method_instruction_evaluator.h"
#include "ar_list.h"
#include "ar_frame.h"
#include <assert.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <inttypes.h>


/**
 * Internal structure for instruction evaluator
 */
struct ar_instruction_evaluator_s {
    ar_log_t *ref_log;                           /* Log instance (borrowed reference) */
    ar_expression_evaluator_t *ref_expr_evaluator;  /* Expression evaluator (borrowed reference) */
    ar_assignment_instruction_evaluator_t *own_assignment_evaluator;  /* Assignment evaluator instance (owned) */
    ar_send_instruction_evaluator_t *own_send_evaluator;  /* Send evaluator instance (owned) */
    ar_condition_instruction_evaluator_t *own_condition_evaluator;  /* Condition evaluator instance (owned) */
    ar_parse_instruction_evaluator_t *own_parse_evaluator;  /* Parse evaluator instance (owned) */
    ar_build_instruction_evaluator_t *own_build_evaluator;  /* Build evaluator instance (owned) */
    ar_method_instruction_evaluator_t *own_method_evaluator;  /* Method evaluator instance (owned) */
    ar_agent_instruction_evaluator_t *own_agent_evaluator;  /* Agent evaluator instance (owned) */
    ar_destroy_agent_instruction_evaluator_t *own_destroy_agent_evaluator;  /* Destroy agent evaluator instance (owned) */
    ar_destroy_method_instruction_evaluator_t *own_destroy_method_evaluator;  /* Destroy method evaluator instance (owned) */
};

/**
 * Creates a new instruction evaluator
 */
ar_instruction_evaluator_t* ar_instruction_evaluator__create(
    ar_log_t *ref_log,
    ar_expression_evaluator_t *ref_expr_evaluator
) {
    // Validate required parameters
    if (ref_log == NULL || ref_expr_evaluator == NULL) {
        return NULL;
    }
    
    // Allocate evaluator structure
    ar_instruction_evaluator_t *evaluator = AR__HEAP__MALLOC(sizeof(ar_instruction_evaluator_t), "instruction_evaluator");
    if (evaluator == NULL) {
        return NULL;
    }
    
    // Initialize fields
    evaluator->ref_log = ref_log;
    evaluator->ref_expr_evaluator = ref_expr_evaluator;
    
    // Create assignment evaluator instance (now uses frame-based pattern)
    evaluator->own_assignment_evaluator = ar_assignment_instruction_evaluator__create(
        ref_log,
        ref_expr_evaluator
    );
    if (evaluator->own_assignment_evaluator == NULL) {
        AR__HEAP__FREE(evaluator);
        return NULL;
    }
    
    // Create send evaluator instance (now uses frame-based pattern)
    evaluator->own_send_evaluator = ar_send_instruction_evaluator__create(
        ref_log,
        ref_expr_evaluator
    );
    if (evaluator->own_send_evaluator == NULL) {
        ar_assignment_instruction_evaluator__destroy(evaluator->own_assignment_evaluator);
        AR__HEAP__FREE(evaluator);
        return NULL;
    }
    
    // Create condition evaluator instance (now uses frame-based pattern)
    evaluator->own_condition_evaluator = ar_condition_instruction_evaluator__create(
        ref_log,
        ref_expr_evaluator
    );
    if (evaluator->own_condition_evaluator == NULL) {
        ar_send_instruction_evaluator__destroy(evaluator->own_send_evaluator);
        ar_assignment_instruction_evaluator__destroy(evaluator->own_assignment_evaluator);
        AR__HEAP__FREE(evaluator);
        return NULL;
    }
    
    // Create parse evaluator instance (now uses frame-based pattern)
    evaluator->own_parse_evaluator = ar_parse_instruction_evaluator__create(
        ref_log,
        ref_expr_evaluator
    );
    if (evaluator->own_parse_evaluator == NULL) {
        ar_condition_instruction_evaluator__destroy(evaluator->own_condition_evaluator);
        ar_send_instruction_evaluator__destroy(evaluator->own_send_evaluator);
        ar_assignment_instruction_evaluator__destroy(evaluator->own_assignment_evaluator);
        AR__HEAP__FREE(evaluator);
        return NULL;
    }
    
    // Create build evaluator instance (now uses frame-based pattern)
    evaluator->own_build_evaluator = ar_build_instruction_evaluator__create(
        ref_log,
        ref_expr_evaluator
    );
    if (evaluator->own_build_evaluator == NULL) {
        ar_parse_instruction_evaluator__destroy(evaluator->own_parse_evaluator);
        ar_condition_instruction_evaluator__destroy(evaluator->own_condition_evaluator);
        ar_send_instruction_evaluator__destroy(evaluator->own_send_evaluator);
        ar_assignment_instruction_evaluator__destroy(evaluator->own_assignment_evaluator);
        AR__HEAP__FREE(evaluator);
        return NULL;
    }
    evaluator->own_method_evaluator = NULL;
    evaluator->own_agent_evaluator = NULL;
    evaluator->own_destroy_agent_evaluator = NULL;
    evaluator->own_destroy_method_evaluator = NULL;
    
    return evaluator;
}

/**
 * Destroys an instruction evaluator
 */
void ar_instruction_evaluator__destroy(ar_instruction_evaluator_t *own_evaluator) {
    if (own_evaluator == NULL) {
        return;
    }
    
    // Destroy owned evaluator instances
    if (own_evaluator->own_assignment_evaluator != NULL) {
        ar_assignment_instruction_evaluator__destroy(own_evaluator->own_assignment_evaluator);
    }
    if (own_evaluator->own_send_evaluator != NULL) {
        ar_send_instruction_evaluator__destroy(own_evaluator->own_send_evaluator);
    }
    if (own_evaluator->own_condition_evaluator != NULL) {
        ar_condition_instruction_evaluator__destroy(own_evaluator->own_condition_evaluator);
    }
    if (own_evaluator->own_parse_evaluator != NULL) {
        ar_parse_instruction_evaluator__destroy(own_evaluator->own_parse_evaluator);
    }
    if (own_evaluator->own_build_evaluator != NULL) {
        ar_build_instruction_evaluator__destroy(own_evaluator->own_build_evaluator);
    }
    if (own_evaluator->own_method_evaluator != NULL) {
        ar_method_instruction_evaluator__destroy(own_evaluator->own_method_evaluator);
    }
    if (own_evaluator->own_agent_evaluator != NULL) {
        ar_agent_instruction_evaluator__destroy(own_evaluator->own_agent_evaluator);
    }
    if (own_evaluator->own_destroy_agent_evaluator != NULL) {
        ar_destroy_agent_instruction_evaluator__destroy(own_evaluator->own_destroy_agent_evaluator);
    }
    if (own_evaluator->own_destroy_method_evaluator != NULL) {
        ar_destroy_method_instruction_evaluator__destroy(own_evaluator->own_destroy_method_evaluator);
    }
    
    // Free the evaluator structure
    AR__HEAP__FREE(own_evaluator);
}



/**
 * Evaluates any instruction AST node
 */
bool ar_instruction_evaluator__evaluate(
    ar_instruction_evaluator_t *mut_evaluator,
    const ar_frame_t *ref_frame,
    const ar_instruction_ast_t *ref_ast
) {
    if (!mut_evaluator || !ref_frame || !ref_ast) {
        return false;
    }
    
    // Dispatch based on AST node type
    ar_instruction_ast_type_t type = ar_instruction_ast__get_type(ref_ast);
    
    switch (type) {
        case AR_INSTRUCTION_AST_TYPE__ASSIGNMENT:
            // Delegate to the assignment instruction evaluator instance (with frame)
            return ar_assignment_instruction_evaluator__evaluate(
                mut_evaluator->own_assignment_evaluator,
                ref_frame,
                ref_ast
            );
            
        case AR_INSTRUCTION_AST_TYPE__SEND:
            // Delegate to the send instruction evaluator instance
            return ar_send_instruction_evaluator__evaluate(
                mut_evaluator->own_send_evaluator,
                ref_frame,
                ref_ast
            );
            
        case AR_INSTRUCTION_AST_TYPE__IF:
            // Delegate to the condition instruction evaluator instance (with frame)
            return ar_condition_instruction_evaluator__evaluate(
                mut_evaluator->own_condition_evaluator,
                ref_frame,
                ref_ast
            );
            
        case AR_INSTRUCTION_AST_TYPE__PARSE:
            // Delegate to the parse instruction evaluator instance
            return ar_parse_instruction_evaluator__evaluate(
                mut_evaluator->own_parse_evaluator,
                ref_frame,
                ref_ast
            );
            
        case AR_INSTRUCTION_AST_TYPE__BUILD:
            // Delegate to the build instruction evaluator instance (with frame)
            return ar_build_instruction_evaluator__evaluate(
                mut_evaluator->own_build_evaluator,
                ref_frame,
                ref_ast
            );
            
        case AR_INSTRUCTION_AST_TYPE__METHOD:
            // Create method evaluator on-demand if needed
            if (mut_evaluator->own_method_evaluator == NULL) {
                ar_data_t *memory = ar_frame__get_memory(ref_frame);
                mut_evaluator->own_method_evaluator = ar_method_instruction_evaluator__create(
                    mut_evaluator->ref_log,
                    mut_evaluator->ref_expr_evaluator,
                    memory
                );
                if (mut_evaluator->own_method_evaluator == NULL) {
                    return false;
                }
            }
            // Delegate to the method instruction evaluator instance
            return ar_method_instruction_evaluator__evaluate(
                mut_evaluator->own_method_evaluator,
                ref_ast
            );
            
        case AR_INSTRUCTION_AST_TYPE__AGENT:
            // Create agent evaluator on-demand if needed
            if (mut_evaluator->own_agent_evaluator == NULL) {
                ar_data_t *memory = ar_frame__get_memory(ref_frame);
                mut_evaluator->own_agent_evaluator = ar_agent_instruction_evaluator__create(
                    mut_evaluator->ref_log,
                    mut_evaluator->ref_expr_evaluator,
                    memory
                );
                if (mut_evaluator->own_agent_evaluator == NULL) {
                    return false;
                }
            }
            // Delegate to the agent instruction evaluator instance
            const ar_data_t *context = ar_frame__get_context(ref_frame);
            return ar_agent_instruction_evaluator__evaluate(
                mut_evaluator->own_agent_evaluator,
                context,
                ref_ast
            );
            
        case AR_INSTRUCTION_AST_TYPE__DESTROY_AGENT:
            // Create destroy agent evaluator on-demand if needed
            if (mut_evaluator->own_destroy_agent_evaluator == NULL) {
                ar_data_t *memory = ar_frame__get_memory(ref_frame);
                mut_evaluator->own_destroy_agent_evaluator = ar_destroy_agent_instruction_evaluator__create(
                    mut_evaluator->ref_log,
                    mut_evaluator->ref_expr_evaluator,
                    memory
                );
                if (mut_evaluator->own_destroy_agent_evaluator == NULL) {
                    return false;
                }
            }
            // Delegate directly to destroy agent evaluator
            return ar_destroy_agent_instruction_evaluator__evaluate(
                mut_evaluator->own_destroy_agent_evaluator,
                ref_ast
            );
            
        case AR_INSTRUCTION_AST_TYPE__DESTROY_METHOD:
            // Create destroy method evaluator on-demand if needed
            if (mut_evaluator->own_destroy_method_evaluator == NULL) {
                ar_data_t *memory = ar_frame__get_memory(ref_frame);
                mut_evaluator->own_destroy_method_evaluator = ar_destroy_method_instruction_evaluator__create(
                    mut_evaluator->ref_log,
                    mut_evaluator->ref_expr_evaluator,
                    memory
                );
                if (mut_evaluator->own_destroy_method_evaluator == NULL) {
                    return false;
                }
            }
            // Delegate directly to destroy method evaluator
            return ar_destroy_method_instruction_evaluator__evaluate(
                mut_evaluator->own_destroy_method_evaluator,
                ref_ast
            );
            
        default:
            // Unknown instruction type
            return false;
    }
}

File: modules/ar_instruction_evaluator.zig
const std = @import("std");
const c = @cImport({
    @cInclude("ar_log.h");
    @cInclude("ar_frame.h");
    @cInclude("ar_instruction_ast.h");
    @cInclude("ar_expression_evaluator.h");
    @cInclude("ar_assignment_instruction_evaluator.h");
    @cInclude("ar_send_instruction_evaluator.h");
    @cInclude("ar_condition_instruction_evaluator.h");
    @cInclude("ar_parse_instruction_evaluator.h");
    @cInclude("ar_build_instruction_evaluator.h");
    @cInclude("ar_compile_instruction_evaluator.h");
    @cInclude("ar_spawn_instruction_evaluator.h");
    @cInclude("ar_exit_instruction_evaluator.h");
    @cInclude("ar_deprecate_instruction_evaluator.h");
    @cInclude("ar_agency.h");
    @cInclude("ar_methodology.h");
});
const ar_allocator = @import("ar_allocator.zig");

/// Internal structure for instruction evaluator facade
const ar_instruction_evaluator_t = struct {
    ref_log: ?*c.ar_log_t,                                              // Log instance (borrowed reference)
    ref_agency: ?*c.ar_agency_t,                                        // Agency instance (borrowed reference)
    own_expr_evaluator: ?*c.ar_expression_evaluator_t,                  // Expression evaluator (owned)
    own_assignment_evaluator: ?*c.ar_assignment_instruction_evaluator_t, // Assignment evaluator (owned)
    own_send_evaluator: ?*c.ar_send_instruction_evaluator_t,            // Send evaluator (owned)
    own_condition_evaluator: ?*c.ar_condition_instruction_evaluator_t,  // Condition evaluator (owned)
    own_parse_evaluator: ?*c.ar_parse_instruction_evaluator_t,          // Parse evaluator (owned)
    own_build_evaluator: ?*c.ar_build_instruction_evaluator_t,          // Build evaluator (owned)
    own_compile_evaluator: ?*c.ar_compile_instruction_evaluator_t,      // Compile evaluator (owned)
    own_spawn_evaluator: ?*c.ar_spawn_instruction_evaluator_t,          // Spawn evaluator (owned)
    own_exit_evaluator: ?*c.ar_exit_instruction_evaluator_t,            // Exit evaluator (owned)
    own_deprecate_evaluator: ?*c.ar_deprecate_instruction_evaluator_t,  // Deprecate evaluator (owned)
};

/// Private implementation that uses error unions for proper cleanup
fn _create(ref_log: ?*c.ar_log_t, ref_agency: ?*c.ar_agency_t) !*ar_instruction_evaluator_t {
    if (ref_log == null or ref_agency == null) return error.NullParameter;
    
    const own_evaluator = ar_allocator.create(ar_instruction_evaluator_t, "instruction_evaluator") orelse 
        return error.OutOfMemory;
    errdefer ar_allocator.free(own_evaluator);
    
    own_evaluator.ref_log = ref_log;
    own_evaluator.ref_agency = ref_agency;
    
    // Create the expression evaluator internally
    own_evaluator.own_expr_evaluator = c.ar_expression_evaluator__create(ref_log) orelse 
        return error.ExpressionEvaluatorCreationFailed;
    errdefer c.ar_expression_evaluator__destroy(own_evaluator.own_expr_evaluator);
    
    // Create all instruction evaluators
    own_evaluator.own_assignment_evaluator = c.ar_assignment_instruction_evaluator__create(
        ref_log,
        own_evaluator.own_expr_evaluator
    ) orelse return error.AssignmentEvaluatorCreationFailed;
    errdefer c.ar_assignment_instruction_evaluator__destroy(own_evaluator.own_assignment_evaluator);
    
    own_evaluator.own_send_evaluator = c.ar_send_instruction_evaluator__create(
        ref_log,
        own_evaluator.own_expr_evaluator
    ) orelse return error.SendEvaluatorCreationFailed;
    errdefer c.ar_send_instruction_evaluator__destroy(own_evaluator.own_send_evaluator);
    
    own_evaluator.own_condition_evaluator = c.ar_condition_instruction_evaluator__create(
        ref_log,
        own_evaluator.own_expr_evaluator
    ) orelse return error.ConditionEvaluatorCreationFailed;
    errdefer c.ar_condition_instruction_evaluator__destroy(own_evaluator.own_condition_evaluator);
    
    own_evaluator.own_parse_evaluator = c.ar_parse_instruction_evaluator__create(
        ref_log,
        own_evaluator.own_expr_evaluator
    ) orelse return error.ParseEvaluatorCreationFailed;
    errdefer c.ar_parse_instruction_evaluator__destroy(own_evaluator.own_parse_evaluator);
    
    own_evaluator.own_build_evaluator = c.ar_build_instruction_evaluator__create(
        ref_log,
        own_evaluator.own_expr_evaluator
    ) orelse return error.BuildEvaluatorCreationFailed;
    errdefer c.ar_build_instruction_evaluator__destroy(own_evaluator.own_build_evaluator);
    
    // Get methodology from agency
    const ref_methodology = c.ar_agency__get_methodology(ref_agency);
    if (ref_methodology == null) return error.MethodologyNotFound;
    
    own_evaluator.own_compile_evaluator = c.ar_compile_instruction_evaluator__create(
        ref_log,
        own_evaluator.own_expr_evaluator,
        ref_methodology
    ) orelse return error.CompileEvaluatorCreationFailed;
    errdefer c.ar_compile_instruction_evaluator__destroy(own_evaluator.own_compile_evaluator);
    
    own_evaluator.own_spawn_evaluator = c.ar_spawn_instruction_evaluator__create(
        ref_log,
        own_evaluator.own_expr_evaluator,
        ref_agency
    ) orelse return error.SpawnEvaluatorCreationFailed;
    errdefer c.ar_spawn_instruction_evaluator__destroy(own_evaluator.own_spawn_evaluator);
    
    own_evaluator.own_exit_evaluator = c.ar_exit_instruction_evaluator__create(
        ref_log,
        own_evaluator.own_expr_evaluator,
        ref_agency
    ) orelse return error.ExitEvaluatorCreationFailed;
    errdefer c.ar_exit_instruction_evaluator__destroy(own_evaluator.own_exit_evaluator);
    
    own_evaluator.own_deprecate_evaluator = c.ar_deprecate_instruction_evaluator__create(
        ref_log,
        own_evaluator.own_expr_evaluator,
        ref_methodology
    ) orelse return error.DeprecateEvaluatorCreationFailed;
    
    return own_evaluator;
}

/// Creates a new instruction evaluator facade (ABI-compatible)
export fn ar_instruction_evaluator__create(
    ref_log: ?*c.ar_log_t,
    ref_agency: ?*c.ar_agency_t
) ?*ar_instruction_evaluator_t {
    return _create(ref_log, ref_agency) catch null;
}

/// Destroys an instruction evaluator facade
export fn ar_instruction_evaluator__destroy(
    own_evaluator: ?*ar_instruction_evaluator_t
) void {
    if (own_evaluator == null) {
        return;
    }
    
    // Destroy owned evaluator instances in reverse order
    // All destroy functions handle null gracefully
    c.ar_deprecate_instruction_evaluator__destroy(own_evaluator.?.own_deprecate_evaluator);
    c.ar_exit_instruction_evaluator__destroy(own_evaluator.?.own_exit_evaluator);
    c.ar_spawn_instruction_evaluator__destroy(own_evaluator.?.own_spawn_evaluator);
    c.ar_compile_instruction_evaluator__destroy(own_evaluator.?.own_compile_evaluator);
    c.ar_build_instruction_evaluator__destroy(own_evaluator.?.own_build_evaluator);
    c.ar_parse_instruction_evaluator__destroy(own_evaluator.?.own_parse_evaluator);
    c.ar_condition_instruction_evaluator__destroy(own_evaluator.?.own_condition_evaluator);
    c.ar_send_instruction_evaluator__destroy(own_evaluator.?.own_send_evaluator);
    c.ar_assignment_instruction_evaluator__destroy(own_evaluator.?.own_assignment_evaluator);
    
    // Destroy the owned expression evaluator
    c.ar_expression_evaluator__destroy(own_evaluator.?.own_expr_evaluator);
    
    // Free the evaluator structure
    ar_allocator.free(own_evaluator);
}

/// Evaluates any instruction AST node - pure facade delegation
export fn ar_instruction_evaluator__evaluate(
    ref_evaluator: ?*const ar_instruction_evaluator_t,
    ref_frame: ?*const c.ar_frame_t,
    ref_ast: ?*const c.ar_instruction_ast_t
) bool {
    if (ref_evaluator == null or ref_frame == null or ref_ast == null) {
        return false;
    }
    
    // Dispatch based on AST node type
    const ast_type = c.ar_instruction_ast__get_type(ref_ast);
    
    switch (ast_type) {
        c.AR_INSTRUCTION_AST_TYPE__ASSIGNMENT => {
            // Delegate to the assignment instruction evaluator
            return c.ar_assignment_instruction_evaluator__evaluate(
                ref_evaluator.?.own_assignment_evaluator,
                ref_frame,
                ref_ast
            );
        },
        
        c.AR_INSTRUCTION_AST_TYPE__SEND => {
            // Delegate to the send instruction evaluator
            return c.ar_send_instruction_evaluator__evaluate(
                ref_evaluator.?.own_send_evaluator,
                ref_frame,
                ref_ast
            );
        },
        
        c.AR_INSTRUCTION_AST_TYPE__IF => {
            // Delegate to the condition instruction evaluator
            return c.ar_condition_instruction_evaluator__evaluate(
                ref_evaluator.?.own_condition_evaluator,
                ref_frame,
                ref_ast
            );
        },
        
        c.AR_INSTRUCTION_AST_TYPE__PARSE => {
            // Delegate to the parse instruction evaluator
            return c.ar_parse_instruction_evaluator__evaluate(
                ref_evaluator.?.own_parse_evaluator,
                ref_frame,
                ref_ast
            );
        },
        
        c.AR_INSTRUCTION_AST_TYPE__BUILD => {
            // Delegate to the build instruction evaluator
            return c.ar_build_instruction_evaluator__evaluate(
                ref_evaluator.?.own_build_evaluator,
                ref_frame,
                ref_ast
            );
        },
        
        c.AR_INSTRUCTION_AST_TYPE__COMPILE => {
            // Delegate to the compile instruction evaluator
            return c.ar_compile_instruction_evaluator__evaluate(
                ref_evaluator.?.own_compile_evaluator,
                ref_frame,
                ref_ast
            );
        },
        
        c.AR_INSTRUCTION_AST_TYPE__SPAWN => {
            // Delegate to the spawn instruction evaluator
            return c.ar_spawn_instruction_evaluator__evaluate(
                ref_evaluator.?.own_spawn_evaluator,
                ref_frame,
                ref_ast
            );
        },
        
        c.AR_INSTRUCTION_AST_TYPE__EXIT => {
            // Delegate to the exit instruction evaluator
            return c.ar_exit_instruction_evaluator__evaluate(
                ref_evaluator.?.own_exit_evaluator,
                ref_frame,
                ref_ast
            );
        },
        
        c.AR_INSTRUCTION_AST_TYPE__DEPRECATE => {
            // Delegate to the deprecate instruction evaluator
            return c.ar_deprecate_instruction_evaluator__evaluate(
                ref_evaluator.?.own_deprecate_evaluator,
                ref_frame,
                ref_ast
            );
        },
        
        else => {
            // Unknown instruction type
            return false;
        },
    }
}